sotodlib/io/hk_utils.py

```diff
@@ -0,0 +1,133 @@
+import numpy as np
+import yaml
+import itertools
+
+from so3g.hk import load_range
+from sotodlib import core
+
+
+def _get_swap_dict(d):
+    """
+    Swap dictionary keys and values; useful for organizing fields in
+    config file by device name instead of alias in get_grouped_hkdata().
+    """
+    return {v: k for k, v in d.items()}
+
+
+def get_grouped_hkdata(start, stop, config):
+    """
+    Takes output from load_range(), reconfigures load_range() dictionary
+    slightly. Groups HK field data by corresponding device/feed. Outputs
+    that list of data.
+    """
+    # call load_range()
+    print("running load_range()")
+    hkdata = load_range(start, stop, config=config)
+
+    # load all fields from config file
+    with open(config, 'r') as file:
+        hkconfig = yaml.safe_load(file)
+
+    field_dict = hkconfig['field_list']
+
+    # output a dict of field name, alias, timestamps, and data
+    # for fields that are online given start, stop and keys from load_range()
+    hk_aliases = list(hkdata.keys())
+
+    online_fields = {}
+    for alias in hk_aliases:
+        field_info = {alias: field_dict[alias]}
+        online_fields.update(field_info)
+
+    fields_data = {}
+    for alias in online_fields:
+        time = hkdata[alias][0]
+        data = hkdata[alias][1]
+        field = online_fields[alias]
+
+        info = {field: [alias, time, data]}
+        fields_data.update(info)
+
+    # now group field names by feed/device
+    sorted_fields = sorted(_get_swap_dict(online_fields))
+
+    grouped_feeds = []
+    key_func = lambda field: field.split('.')[0:4]
+    for key, group in itertools.groupby(sorted_fields, key_func):
+        grouped_feeds.append(list(group))
+
+    # use grouped_feeds and fields_data to output a grouped list of data per
+    # feed/device, ready to input to an axismanager
+    grouped_data = []
+    for group in grouped_feeds:
+        device_data = {}
+        for field in group:
+            field_dict = {field: fields_data[field]}
+            device_data.update(field_dict)
+
+        grouped_data.append(device_data)
+
+    return grouped_data
+
+
+def make_hkaman(grouped_data):
+    """
+    Takes data from get_grouped_hkdata(), tests whether feed/device is
+    cosampled, outputs axismanager(s) for either case.
+    """
+    amans = []
+    for group in grouped_data:
+        data = {}
+        aliases = []
+        times = []
+        for field in group:
+            # arrange data per device to input into aman
+            alias = group[field][0]
+            aliases.append(alias)
+
+            time = group[field][1]
+            times.append(time)
+
+            device_data = group[field][2]
+
+            info = {alias: device_data}
+            data.update(info)
+
+        try:
+            # check whether hk device is cosampled
+            assert len(times[0]) == len(times[-1])
+            # TODO: diff of times, if less than say 10%??, treat as cosampled
```

Comment:
times is a list of arrays of times? It can have more than length 2? If so, I think I'd rewrite this to 
```
assert np.all( [len(t)==len(times[0]) for t in times])
```

I do think it's a good idea to also check the samples too with some level of precision. I would make the cosampled precision a kwarg of the function and set it at that 10% guess you started with. That will let folks mess with it later if they want

Reply:
half-resolved! Still need to add the 10% guess; do we need this portion as a first pass for merging? Probably we want UCSD to test this hk tool soon? Wondering if I should make checking the samples a GitHub issue to address as UCSD concurrently tests this tool. What do you think?